alf/algorithms/misc_algorithm.py

```diff
@@ -0,0 +1,263 @@
+        layery = self._misc_layery(inputs=y_conc)
+        layer2 = tf.nn.relu(layerx + layery)
+        output = self._misc_layero(inputs=layer2)
+        output = tf.nn.tanh(output)
+
+        # split in T_xy and T_x_y predictions
+        N_samples = tf.shape(x_in)[1]
+        T_xy = output[:, :N_samples, :]
+        T_x_y = output[:, N_samples:, :]
+
+        # compute the negative loss (maximize loss == minimize -loss)
+        mean_exp_T_x_y = tf.reduce_mean(tf.math.exp(T_x_y), axis=1)
+        loss = tf.reduce_mean(T_xy, axis=1) - tf.math.log(mean_exp_T_x_y)
+        loss = tf.squeeze(loss)  # Mutual Information
+
+        return loss
+
+    def train_step(self, inputs, state, calc_intrinsic_reward=True):
+        """
+        Args:
+            inputs (tuple): observation
+            state (Tensor):  state for MISC (previous feature)
+            calc_intrinsic_reward (bool): if False, only return the losses
+        Returns:
+            TrainStep:
+                outputs: empty tuple ()
+                state:  empty tuple ()
+                info: (MISCInfo):
+        """
+        feature_state, prev_action = inputs
+        feature = tf.concat([feature_state, prev_action], axis=-1)
+        prev_feature = state
+
+        feature_reshaped = tf.expand_dims(feature, axis=1)
+        prev_feature_reshaped = tf.expand_dims(prev_feature, axis=1)
+        feature_pair = tf.concat([prev_feature_reshaped, feature_reshaped], 1)
+        feature_reshaped_tran = tf.transpose(feature_reshaped, perm=[1, 0, 2])
+
+        def add_batch():
+            self._buffer.add_batch(feature_reshaped_tran)
+
+        # run add_batch function, if there are new trajectories being rollouted
+        run_if(
+            tf.reduce_all(
+                tf.equal(
+                    tf.shape(feature),
+                    tf.constant(self._traj_spec.shape.as_list()))), add_batch)
+
+        if self._n_objects < 2:
+            obs_tau_excludes_goal, obs_tau_achieved_goal = self.split_observation_tf(
+                feature_pair)
+            loss = self.mine(obs_tau_excludes_goal, obs_tau_achieved_goal)
+        elif self._n_objects == 2:
+            obs_tau_excludes_goal, obs_tau_achieved_goal_1, obs_tau_achieved_goal_2 = self.split_observation_tf(
+                feature_pair)
+            loss_1 = self.mine(obs_tau_excludes_goal, obs_tau_achieved_goal_1)
+            loss_2 = self.mine(obs_tau_excludes_goal, obs_tau_achieved_goal_2)
+            loss = loss_1 + loss_2
+
+        intrinsic_reward = ()
+        if calc_intrinsic_reward:
+            # scale/normalize the MISC intrinsic reward
+            if self._n_objects < 2:
+                intrinsic_reward = tf.clip_by_value(self._mi_r_scale * loss,
+                                                    *(0, 1))
+            elif self._n_objects == 2:
+                intrinsic_reward = tf.clip_by_value(
+                    self._mi_r_scale * loss_1, *(0, 1)) + 1 * tf.clip_by_value(
```

Comment:
why do you need `*(0, 1)` instead of `0, 1`?
why do you need `1 *`?

Reply:
? what do you mean?

Reply:
use as required